lib/db/util.go

```diff
@@ -17,11 +17,26 @@ const (
 // FileInfoBatch is a utility to do file operations on the database in suitably
 // sized batches.
 type FileInfoBatch struct {
-	infos   []protocol.FileInfo
-	size    int
-	flushFn func([]protocol.FileInfo) error
+	infos      []protocol.FileInfo
+	size       int
+	flushFn    func([]protocol.FileInfo) error
+	reuseSlice bool
 }
 
+// NewReusingFileInfoBatch returns a new FileInfoBatch that calls fn when
+// it's time to flush. The given slice of FileInfos is a view into the
+// internal buffer which will be reused, so it must not be read after
+// returning from the flush function.
+func NewReusingFileInfoBatch(fn func([]protocol.FileInfo) error) *FileInfoBatch {
```

Comment:
Do we want this at all? In most (maybe all) usages we pass the slice to the db layer. I am convinced the current one does everything synchronously and serialises things, so no references are kept after the db layer returns. It's however not given that's the case - e.g. one could imagine the db caching/batching by keeping the data in the slice. Doesn't sound convincing/realistic really - still my point is more if having to reason about potential issues is worth the performance/allocation benefit?

Reply:
Maybe not? There's the question of context as well. Saving a slice allocation in the hot path is all good and well, but if the next step is to pass it to the database where it will get serialized, copied, compressed, hashed, written to disk, etc I doubt it matters at all.